client/signup/steps/page-picker/use-cart-for-difm.tsx

```diff
@@ -0,0 +1,321 @@
+			itemSubTotal: 0,
+			// The 0 selection product cost should be shown as 0
+			productCost: 0,
+			nameOverride: translate( '0 Extra Pages' ),
+			meta: translate( '%(perPageCost)s Per Page', {
+				args: {
+					perPageCost: formatCurrency( extraPageProduct.cost, currencyCode, { precision: 0 } ),
+				},
+			} ),
+		} );
+	}
+	return cartItems;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const debounce = ( callback: ( ...args: any[] ) => any, timeout: number ) => {
+	let timeoutId: number | undefined = undefined;
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	return ( ...args: any[] ) => {
+		window.clearTimeout( timeoutId );
+		timeoutId = window.setTimeout( () => {
+			callback( ...args );
+		}, timeout );
+	};
+};
+
+export function useCartForDIFM( selectedPages: string[] ): {
+	items: CartItem[];
+	total: string | null;
+	isCartLoading: boolean;
+	isCartPendingUpdate: boolean;
+	isCartUpdateStarted: boolean;
+	isProductsLoading: boolean;
+	effectiveCurrencyCode: string | null;
+} {
+	//This state is used by loader states to provide immediate feedback when a deebounced change happens to a cart
+	const [ isCartUpdateStarted, setIsCartUpdateStarted ] = useState( false );
+
+	const signupDependencies = useSelector( getSignupDependencyStore );
+	const translate = useTranslate();
+	const dispatch = useDispatch();
+	const proPlan = useSelector( ( state ) => getProductBySlug( state, PLAN_WPCOM_PRO ) );
+	const premiumPlan = useSelector( ( state ) => getProductBySlug( state, PLAN_PREMIUM ) );
+	const activePremiumPlanScheme = isEnabled( 'plans/pro-plan' ) ? proPlan : premiumPlan;
+	const { newOrExistingSiteChoice, siteId, siteSlug } = signupDependencies;
+
+	const isProductsLoading = useSelector( isProductsListFetching );
+	const extraPageProduct = useSelector( ( state ) =>
+		getProductBySlug( state, WPCOM_DIFM_EXTRA_PAGE )
+	);
+	const difmLiteProduct = useSelector( ( state ) => getProductBySlug( state, WPCOM_DIFM_LITE ) );
+
+	const userCurrencyCode = useSelector( getCurrentUserCurrencyCode );
+
+	const site = useSelector( ( state ) => getSite( state, siteSlug ?? siteId ) );
+	const cartKey = getCartKey( { selectedSite: site } );
+
+	const { replaceProductsInCart, responseCart, isLoading, isPendingUpdate } = useShoppingCart(
+		cartKey ?? undefined
+	);
+	const getEffectiveCurrencyCode = useCallback( () => {
+		if ( newOrExistingSiteChoice === 'existing-site' ) {
+			return responseCart.currency ?? userCurrencyCode;
+		}
+		return userCurrencyCode;
+	}, [ newOrExistingSiteChoice, responseCart.currency, userCurrencyCode ] );
+
+	const getDifmLiteCartProduct = useCallback( () => {
+		if ( difmLiteProduct ) {
+			return {
+				...difmLiteProduct,
+				extra: buildDIFMCartExtrasObject( {
+					...signupDependencies,
+					selectedPageTitles: selectedPages,
+				} ),
+			};
+		}
+		return null;
+	}, [ signupDependencies, selectedPages, difmLiteProduct ] );
```

Comment:
But the DIFM lite will have to be passed via the use effect. And now the useEffect will fire each time page changes. The main objective of pushing this section out of the effect was to mainly reduce the number of dependencies on that effect. Beyond that the level of optimisation remains more or less the same same IMO since this is merely a simple object creation.

I feel like having the cart product creation logic encapsulated at this point a bigger win. But let me know if you think otherwise :)